cunumeric/deferred.py

```diff
@@ -974,23 +831,92 @@ def contract(
         rhs2_modes,
         mode2extent,
     ):
+        supported_dtypes = [
+            np.float16,
+            np.float32,
+            np.float64,
+            np.complex64,
+            np.complex128,
+        ]
         lhs_thunk = self
 
-        # TODO: More sanity checks (no duplicate modes, no singleton modes, no
-        # broadcasting, ...)
-
-        # Casting should have been handled by the frontend
-        assert lhs_thunk.dtype is rhs1_thunk.dtype
-        assert lhs_thunk.dtype is rhs2_thunk.dtype
+        # Sanity checks
+        # no duplicate modes within an array
+        assert len(lhs_modes) == len(set(lhs_modes))
+        assert len(rhs1_modes) == len(set(rhs1_modes))
+        assert len(rhs2_modes) == len(set(rhs2_modes))
+        # no singleton modes
+        mode_counts = Counter()
+        mode_counts.update(lhs_modes)
+        mode_counts.update(rhs1_modes)
+        mode_counts.update(rhs2_modes)
+        for count in mode_counts.values():
+            assert count == 2 or count == 3
+        # arrays and mode lists agree on dimensionality
+        assert lhs_thunk.ndim == len(lhs_modes)
+        assert rhs1_thunk.ndim == len(rhs1_modes)
+        assert rhs2_thunk.ndim == len(rhs2_modes)
+        # array shapes agree with mode extents (broadcasting should have been
+        # handled by the frontend)
+        assert all(
```

Comment:
Do we have a sense of how expensive these assertions are? I know we want to check that the frontend correctly filters out unsupported cases, but the amount of checks in this function seems somewhat excessive, especially because a similar amount of checks would be already done in the frontend.

Reply:
I expect mode strings to be short (5-ish characters each), so these checks won't be horrible. You are right that they're all for internal checking (there should be no user input that will trigger these, only bugs in the frontend). How about I only do them if we're running with `-cunumeric:test`?

Reply:
That may be an option, but I start to feel we should have a different flag, say `-cunumeric:debug`, that turns on all debugging checks and assertions.

Reply:
Sounds good, I don't have a strong opinion either way. Should I go ahead and add `-cunumeric:debug` on this PR?

Reply:
Let's make a separate PR for it, as this isn't the only check we want to guard with the debug flag.